### PR TITLE
feat EXTCODECOPY

### PR DESCRIPTION
**Pull request by jobez**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X]  Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

A lack of implementation of EXTCODECOPY opcode.

Issue Number: #95 

## What is the new behavior?

An implementation of EXTCODECOPY opcode.

-
-
-

### CODE REVIEW

File: src/kakarot/instructions/environmental_information.cairo [hunk 1]
```diff
@@ -397,6 +398,88 @@ namespace EnvironmentalInformation {
...
+
+        // Get the starknet address from the given evm address
+        let (registry_address_) = registry_address.read();
+
+        let (starknet_contract_address) = IRegistry.get_starknet_contract_address(
+            contract_address=registry_address_, evm_contract_address=address
+        );
+
+        // handle case where there is no eth -> stark address mapping
+        if (starknet_contract_address == 0) {
```
Comment (Eikix) [inline src/kakarot/instructions/environmental_information.cairo:442]: Does reading a uninitialized storage value in Starknet return a 0?

File: src/kakarot/instructions/environmental_information.cairo [hunk 2]
```diff
@@ -397,6 +398,88 @@ namespace EnvironmentalInformation {
...
+        }
+
+        // Get the bytecode from the Starknet_contract
+        let (bytecode_len, bytecode) = IEvmContract.bytecode(
+            contract_address=starknet_contract_address
+        );
+
+        // handle case were eth address returns no bytecode: 
+        
+        if (bytecode_len == 0) {
```
Comment (Eikix) [inline src/kakarot/instructions/environmental_information.cairo:453]: Same question as above

File: tests/cairo_files/instructions/test_environmental_information.cairo [hunk 3]
```diff
@@ -76,3 +77,33 @@ func test__exec_address__should_push_address_to_stack{
...
+    let stack: model.Stack* = Stack.push(stack, Uint256(3, 0)); // address
+
+    let ctx: model.ExecutionContext* = TestHelpers.init_context_with_stack(0, bytecode, stack);
+
+
+    // When
+    let result = EnvironmentalInformation.exec_extcodecopy(ctx);
+
+    // Then
+    assert result.gas_used = ctx.gas_used;
```
Comment (Eikix) [inline tests/cairo_files/instructions/test_environmental_information.cairo:106]: What does this `assert` test?

File: tests/cairo_files/instructions/test_environmental_information.cairo [hunk 4]
```diff
@@ -76,3 +77,33 @@ func test__exec_address__should_push_address_to_stack{
...
+    let stack: model.Stack* = Stack.push(stack, Uint256(1, 2)); // size
+    let stack: model.Stack* = Stack.push(stack, Uint256(1, 0)); // offset
+    let stack: model.Stack* = Stack.push(stack, Uint256(2, 0)); // dest_offset
+    let stack: model.Stack* = Stack.push(stack, Uint256(3, 0)); // address
+
+    let ctx: model.ExecutionContext* = TestHelpers.init_context_with_stack(0, bytecode, stack);
+
+
+    // When
+    let result = EnvironmentalInformation.exec_extcodecopy(ctx);
```
Comment (Eikix) [inline tests/cairo_files/instructions/test_environmental_information.cairo:103]: What are we testing here?
We'd need an integration test to:
Deploy a contract through Kakarot -> Execute extcodecopy on it

File: src/kakarot/instructions/environmental_information.cairo (hunk 1 above)
Comment (jobez) [reply src/kakarot/instructions/environmental_information.cairo:442]: I believe so, in practice I get this result and also it is mentioned here: https://hackmd.io/@RoboTeddy/BJZFu56wF#Ambiguity-between-uninitialized-memory-and-the-value-0

File: tests/cairo_files/instructions/test_environmental_information.cairo (hunk 4 above)
Comment (jobez) [reply tests/cairo_files/instructions/test_environmental_information.cairo:103]: @Eikix i made this test to stub out a test.

I understand now that a minimal unit test will need to deploy an account registry. Working on that now!

File: src/kakarot/instructions/environmental_information.cairo [hunk 5]
```diff
@@ -397,6 +398,86 @@ namespace EnvironmentalInformation {
...
+        let address = popped[0];
+        let dest_offset = popped[1];
+        let offset = popped[2];
+        let size = popped[3];
+      
+        // Get the starknet address from the given evm address
+        let (registry_address_) = registry_address.read();
+
+        let (starknet_contract_address) = IRegistry.get_starknet_contract_address(
+            contract_address=registry_address_, evm_contract_address=address.low
```
[the remark this answers is not in the document]
Comment (jobez) [reply src/kakarot/instructions/environmental_information.cairo:436]: great point. i ran into this when doing integration testing, thankfully there is a helper function that goes from uint256 to felt that I am using

File: tests/integrations/test_kakarot.py [hunk 6]
```diff
@@ -9,13 +10,38 @@
     traceit,
 )
 
+@pytest.mark.asyncio
+class TestExtCodeCopy:
```
[the remark this answers is not in the document]
Comment (jobez) [reply tests/integrations/test_kakarot.py:14]: will do 🖖

File: src/kakarot/instructions/environmental_information.cairo [hunk 7]
```diff
@@ -424,16 +424,18 @@ namespace EnvironmentalInformation {
         // 2 - offset: byte offset in the code to copy.
         // 3 - size: byte size to copy.
         let (stack, popped) = Stack.pop_n(self=stack, n=4);
-        let address = popped[0];
+        let address_ = popped[0];
```
Comment (ClementWalter) [inline src/kakarot/instructions/environmental_information.cairo:427]: would not block a PR for this but I'd like to have `address_uint256` & `address_felt` or something explicit like this when the two versions are used in the same piece of code. `address_` is less explicit

Comment (jobez) [reply src/kakarot/instructions/environmental_information.cairo:427]: makes sense ⚡

**Review by ClementWalter — DISMISSED** [review]

**Comment by jobez** [comment]
> there's something weird with the commit history and displayed diff, can you rebase / squash maybe?

Sorry about that. Should be better behaved!

**Pull request merged**